### PR TITLE
Kalman filter

### PR DESCRIPTION
**Pull request by thowell**
Builds on #149 

Implements an extended Kalman filter
-includes gRPC + Python interface + tests

Modifications to Estimator class and corresponding updates to Batch estimator

Minor changes to tests

### CODE REVIEW

File: grpc/kalman_service.cc [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+  if (!Initialized()) {
+    return {grpc::StatusCode::FAILED_PRECONDITION, "Init not called."};
+  }
+
+  // unpack input/output
+  kalman::Covariance input = request->covariance();
+  kalman::Covariance* output = response->mutable_covariance();
+
+  // dimensions
+  int nvelocity = 2 * kalman_.model->nv;
```
Comment (thowell) [inline grpc/kalman_service.cc:249]: Needs to be 2 * nv + na

File: grpc/kalman_service.cc [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+    return {grpc::StatusCode::FAILED_PRECONDITION, "Init not called."};
+  }
+
+  // unpack input/output
+  kalman::Noise input = request->noise();
+  kalman::Noise* output = response->mutable_noise();
+
+  // dimensions
+  int nprocess = 2 * kalman_.model->nv;
+  int nsensor = kalman_.model->nsensordata;
```
Comment (thowell) [inline grpc/kalman_service.cc:283]: needs to be kalman_.nsensordata_

File: grpc/kalman_service.cc [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+  if (!Initialized()) {
+    return {grpc::StatusCode::FAILED_PRECONDITION, "Init not called."};
+  }
+
+  // unpack input/output
+  kalman::Noise input = request->noise();
+  kalman::Noise* output = response->mutable_noise();
+
+  // dimensions
+  int nprocess = 2 * kalman_.model->nv;
```
Comment (thowell) [inline grpc/kalman_service.cc:282]: needs to be 2 * nv + na

File: mjpc/estimators/batch.h [hunk 4]
```diff
@@ -402,6 +436,8 @@ class Batch : public Estimator {
   double cost_initial_;
   double cost_previous_;
 
+  // TODO(taylor): underscore
```
Comment (thowell) [inline mjpc/estimators/batch.h:439]: Fixed in subsequent PR

File: mjpc/estimators/estimator.h [hunk 5]
```diff
@@ -72,19 +75,201 @@ class Estimator {
...
+
+  // estimator-specific GUI elements
+  void GUI(mjUI& ui, EstimatorGUIData& data) override{};
+
+  // set GUI data
+  void SetGUIData(EstimatorGUIData& data) override{};
+
+  // estimator-specific plots
+  void Plots(mjvFigure* fig_planner, mjvFigure* fig_timer, int planner_shift,
+             int timer_shift, int planning, int* shift) override{};
```
Comment (thowell) [inline mjpc/estimators/estimator.h:246]: Fix space

File: mjpc/estimators/kalman.cc [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+
+  // correction
+  mju_zero(correction_.data(), ndstate_);
+
+  // timer
+  timer_measurement_ = 0.0;
+  timer_prediction_ = 0.0;
+
+  // scratch
+  std::fill(tmp0_.begin(), tmp0_.end(), 0.0);
```
Comment (thowell) [inline mjpc/estimators/kalman.cc:156]: unify into a single scratch memory

File: python/mujoco_mpc/batch.py [hunk 7]
```diff
@@ -504,7 +504,7 @@ def status_code(code):
       elif code == 3:
         return "SMALL_DIRECTION_FAILURE"
       elif code == 4:
-        return "MAX_REGULARIZATION_FAILURE"
+        return "kMaxBatchRegularization_FAILURE"
```
Comment (thowell) [inline python/mujoco_mpc/batch.py:507]: Fixed in subsequent PR

**Comment by thowell** [comment]
Kalman gRPC + Python is now removed.

**Review by erez-tom — APPROVED** [review]

**Pull request merged**